R-package/src/cmake/modules/FindLibR.cmake

```diff
@@ -0,0 +1,210 @@
+# CMake module used to find the location of R's
+# dll and header files.
+#
+# Borrows heavily from xgboost's R package:
+#
+# * https://github.com/dmlc/xgboost/blob/master/cmake/modules/FindLibR.cmake
+#
+# Defines the following:
+#  LIBR_FOUND
+#  LIBR_HOME
+#  LIBR_EXECUTABLE
+#  LIBR_INCLUDE_DIRS
+#  LIBR_LIB_DIR
+#  LIBR_CORE_LIBRARY
+# and a CMake function to create R.lib for MSVC
+
+if(NOT R_ARCH)
+  if("${CMAKE_SIZEOF_VOID_P}" STREQUAL "4")
+    set(R_ARCH "i386")
+  else()
+    set(R_ARCH "x64")
+  endif()
+endif()
+
+if(NOT ("${R_ARCH}" STREQUAL "x64"))
+  message(FATAL_ERROR "LightGBM's R package currently only supports 64-bit operating systems")
+endif()
+
+# Creates R.lib and R.def in the build directory for linking with MSVC
+function(create_rlib_for_msvc)
+
+  message("Creating R.lib and R.def")
+
+  # various checks and warnings
+  if(NOT WIN32 OR NOT MSVC)
+    message(FATAL_ERROR "create_rlib_for_msvc() can only be used with MSVC")
+  endif()
+
+  if(NOT EXISTS "${LIBR_LIB_DIR}")
+    message(FATAL_ERROR "LIBR_LIB_DIR, '${LIBR_LIB_DIR}', not found")
+  endif()
+
+  find_program(GENDEF_EXE gendef)
+  find_program(DLLTOOL_EXE dlltool)
+
+  if(NOT GENDEF_EXE OR NOT DLLTOOL_EXE)
+    message(FATAL_ERROR "Either gendef.exe or dlltool.exe not found!\
+      \nDo you have Rtools installed with its MinGW's bin/ in PATH?")
+  endif()
+
+  # extract symbols from R.dll into R.def and R.lib import library
+  execute_process(COMMAND ${GENDEF_EXE}
+    "-" "${LIBR_LIB_DIR}/R.dll"
+    OUTPUT_FILE "${CMAKE_CURRENT_BINARY_DIR}/R.def"
+  )
+  execute_process(COMMAND ${DLLTOOL_EXE}
+    "--input-def" "${CMAKE_CURRENT_BINARY_DIR}/R.def"
+    "--output-lib" "${CMAKE_CURRENT_BINARY_DIR}/R.lib"
+  )
+endfunction(create_rlib_for_msvc)
+
+# R version information is used to search for R's libraries in
+# the registry on Windows. Since this code is orchestrated by
+# an R script (src/install.libs.R), that script uses R's built-ins to
+# find the version of R and pass it through as a CMake variable
+if(CMAKE_R_VERSION)
+  message("R version passed into FindLibR.cmake: ${CMAKE_R_VERSION}")
+elseif(WIN32)
+  message(FATAL_ERROR "Expected CMAKE_R_VERSION to be passed in on Windows but none was provided. Check src/install.libs.R")
+endif()
+
+
+if (NOT LIBR_EXECUTABLE)
+  find_program(
+    LIBR_EXECUTABLE
+    NAMES R R.exe
+  )
+
+  # CRAN may run RD CMD CHECK instead of R CMD CHECK,
+  # which can lead to this infamous error:
+  # 'R' should not be used without a path -- see par. 1.6 of the manual
+  if(LIBR_EXECUTABLE MATCHES ".*\\.Rcheck.*")
+    unset(LIBR_EXECUTABLE PARENT_SCOPE)
+    unset(LIBR_EXECUTABLE CACHE)
+  endif()
+
+  # ignore the R bundled with R.app on Mac, since that is GUI-only
+  if(LIBR_EXECUTABLE MATCHES ".+R\\.app.*")
```

Comment:
Shouldn't it end with `$` `".+R\\.app$"` as we search for `R.app` executable?

Reply:
No we're not searching for an executable called `R.app`. This is found because we're searching for an executable called `R` that ends up in a path like `/Applications/R.app/Contents/MacOS/R`. It's regrettable that they called the GUI-only executable `R` (the same as the headless one).

So this check is "if you found something in any path with `R.app` in it

Reply:
Ah, I see! Thanks for the clarification!